### ISSUE TITLE
API Overhaul: Preview Templating

### ISSUE THREAD

**Issue by erquhart** [netlify/netlify-cms#1041]
Let's make custom preview templating better.

This issue outlines a potential approach for improving the custom preview authoring experience through documentation driven development. Rough documentation will be drawn up right in the comments describing an ideal custom preview interface, and POC (proof of concept) validating these proposals can be linked as we go.

### Current
Currently, registering a preview template looks like this:

```js
var PostPreview = createClass({
  render: function() {
    var entry = this.props.entry;
    var image = entry.getIn(['data', 'image']);
    var bg = image && this.props.getAsset(image);
    return h('div', {},
      h('div', {className: "cover"},
        h('h1', {}, entry.getIn(['data', 'title'])),
        bg ? h('img', {src: bg.toString()}) : null
      ),
      h('p', {},
        h('small', {}, "Written " + entry.getIn(['data', 'date']))
      ),
      h('div', {"className": "text"}, this.props.widgetFor('body'))
    );
  }
});
```

It's a generally effective templating system, but there are two primary drawbacks:

* It's written to work with React's `createElement` method, which isn't a newfangled interface by any means, but may be foreign to a lot of folks.
* It's hampered by the awkward `widgetFor`/`widgetsFor` interface for retrieving widgets and entry data.

### Proposed
Ideally, developers could provide a simple mustache/handlebars template - here's the above example rewritten in handlebars:

```js
var postPreviewTemplate = `
  <div>
    <div class="cover">
      <h1>{{title}}</h1>
      {{#image}}
        <img src="{{getAsset image}}"/>
      {{/image}}
    </div>
    <p>
      <small>Written {{date}}</small>
    </p>
    <div class="text">
      {{widgetFor "body"}}
    </div>
  </div>
`;
```

This style of templating should prove far more approachable, simpler for troubleshooting, and clearer in purpose. Handlebars helper functions would be created to apply necessary functionality, for example, our use of `getAsset` in the template above.

### widgetFor / widgetsFor
As mentioned before, the `widgetFor` / `widgetsFor` methods for getting values from shallow and deep fields, respectively, are confusing to use. `widgetFor` accepts a field name and returns a React component with the field's value wrapped in the preview component, ready for rendering. `widgetsFor` accepts an array of keys for accessing a nested field, but instead of returning a React component, it returns an object with keys "widget" and "data", for accessing the component or just the raw value respectively.

Instead, `widgetFor` should handle both jobs, accepting either a field name string, or else an array of nested field names for deep retrieval. It should always return a React component, as raw values, including nested ones, are already available on the `entry` object.

### Current
An implementation from the example project that uses both `widgetFor` and `widgetsFor`:

```js
var GeneralPreview = createClass({
  render: function() {
    var entry = this.props.entry;
    var title = entry.getIn(['data', 'site_title']);
    var posts = entry.getIn(['data', 'posts']);
    var thumb = posts && posts.get('thumb');

    return h('div', {},
      h('h1', {}, title),
      h('dl', {},
        h('dt', {}, 'Posts on Frontpage'),
        h('dd', {}, this.props.widgetFor(['posts', 'front_limit']) || 0),

        h('dt', {}, 'Default Author'),
        h('dd', {}, this.props.widgetsFor('posts').getIn(['data', 'author']) || 'None'),

        h('dt', {}, 'Default Thumbnail'),
        h('dd', {}, thumb && h('img', {src: this.props.getAsset(thumb).toString()}))
      )
    );
  }
});
```

### Proposed
Here's what the above template would look like with the proposed removal of `widgetsFor`:

```js
var generalPreviewTemplate = `
  <div>
    <h1>{{ site_title }}</h1>
    <dl>
      {{#posts}}
        <dt>Posts on Frontpage</dt>
        <dd>{{widgetFor "front_limit"}}</dd>

        <dt>Default Author</dt>
        <dd>{{author}}</dd>

        <dt>Default Thumbnail</dt>
        <dd>
          {{#thumb}}<img src="{{getAsset thumb}}"/>{{/thumb}}
      {{/posts}}
    </dl>
  </div>
`;
```

### Proof of Concept
A branch can be referenced here:
https://github.com/netlify/netlify-cms/compare/api-register-preview-template

Deploy preview here:
https://api-register-preview-template--cms-demo.netlify.com

The first example above is working in the POC - the preview for Post entries is created using a handlebars template in `example/index.html`. The second example has not yet been implemented.

## Documentation
(TBA)

**Comment by biilmann** [comment]
Great stuff!

I agree with getting rid of the `widgetFor` and `widgetsFor` distinction, always felt a bit clunky. I feel a bit the same way around the current metadata construct used when accessing relations content in a custom preview template.

The old ember-prototype of the CMS used handlebars since that's native to ember, and for simple preview templates that's a lot more straight forward to quickly put together than React components. However, I do think the underlying React component concept is more robust and should always be an option.

My suggestion would be that we make a template language abstraction and implement handlebars (or mustache or something with a similar feel) as a built in template language. The abstraction would be something like:

```js
class TemplateLanguage {
   compile(template : String) {
       return compileTemplateIntoReactComponent(template)
   }
}
```

Another nicety for the Ember version was loading templates from script tags. This made it really easy to get the CMS going just by pulling in the cdn lib and then adding some preview templates in the index.html like:

```
<script type="text/x-handlebars-template" name="post">
<div>
  <h1>{{ entry.title }}</h1>

   <div class="text">{{ widgetFor "body" }}</div>
</div>
</script>

**Comment by BinaryBen** [comment]
YES PLEASE! This looks much simpler. My main comment on this is that it needs to handle some basic if/then and loop (for lists) logic for some pages if the proposed handlebars option doesn't already include that.

I'm also wondering if there is any way to re-use existing templates and partials used by popular static site generators? It kind of sucks a bit to have to write them for use on the website, and then have to rewrite/rework them again just to preview in the CMS portal.

**Comment by erquhart** [comment]
Crystal clear: template reuse is the name of the game. I'm thinking we'll add `registerTemplateParser` to allow custom parsers. True template reuse is going to be challenging, but I don't think it's impossible. Even Go templates can be parsed in the browser via Gopherjs.

The challenge is working with a template's data structure, which may be much different from what we have in Netlify CMS. Perhaps a bit of extensible mapping with transforms could stand in the gap. Need to try some of this out with a real project and I'll update the OP with some more concepts from there.

**Comment by erquhart** [comment]
Alrighty, documentation rough draft complete, thoughts welcome. I'll get some more POC work added tomorrow. Considering the value of this documentation, I'm considering doing these API overhaul tickets partially in a PR so things are versioned.

**Comment by tech4him1** [comment]
@erquhart @biilmann Are we planning on "compiling" the templates into React components, then running them? Or just using them as templates directly, with their own engine?

**Comment by erquhart** [comment]
Compiling to React - there's a working handlebars implementation in the POC: https://github.com/netlify/netlify-cms/compare/api-register-preview-template

**Comment by tech4him1** [comment]
Ah, but you're still compiling to React *after* you run it though Handlebars, if I understand correctly.
I was thinking you were compiling to react first, with all variables intact, then parsing them in React.

**Comment by erquhart** [comment]
That's be awesome, but we'd need a lib that does that for each template language - surprisingly, there isn't even one for handlebars, I checked before starting.

**Comment by erquhart** [comment]
Since every parser is unique, we'll need to allow some kind of pass through, like an opts object maybe (but not really options), so that the parser can accept helper registration, partials, whatever it may want to expose.

Going to look at the abstraction layer concept now.

**Comment by haileypate** [comment]
Since I use jinja2 to process my templates and generate site files, I'd love to be able to point netlifyCMS to a directory in which I've prepared templates that could get gobbled up and leveraged by the `EditorPreviewPane` component. 

I figured I'd be doing double work to recreate templates for use in netlifyCMS's preview feature for the forseeable future... but... I'd totally invest in time in leveraging these new options

Do y'all think a node package like [jinja-to-js](https://www.npmjs.com/package/jinja-to-js) might be useful to someone like me in the above workflow?

**Comment by erquhart** [comment]
Let me get further in the docs, I'm pushing up a guide to creating template compilers next, which is what you'd need to do to support Jinja templates.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by mischkl** [comment]
Is this still on track to be implemented sometime? I would love to be able to reuse Go templates from Hugo without in the NetlifyCMS preview pane without having to rewrite them in React.

**Comment by erquhart** [comment]
Here's the reference for Hugo/Go templates in the browser: https://github.com/gohugoio/hugo/issues/4543

Still a ways to go on that, I was overly optimistic when I authored this issue.

**Comment by valeriy-ten-aspose** [comment]
> Here's the reference for Hugo/Go templates in the browser: [gohugoio/hugo#4543](https://github.com/gohugoio/hugo/issues/4543)
> 
> Still a ways to go on that, I was overly optimistic when I authored this issue.

@erquhart is there any progress on that? 
I configured NetlifyCMS for my Hugo site that have markdown files at self-hosted Gitlab. Now I can't make NetlifyCMS preview pane to display pages with Hugo styles and templates. Could you please clarify is this possible at all?

**Comment by ValeriyTen** [comment]
> Here's the reference for Hugo/Go templates in the browser: [gohugoio/hugo#4543](https://github.com/gohugoio/hugo/issues/4543)
> 
> Still a ways to go on that, I was overly optimistic when I authored this issue.

@erquhart is there any progress on that?
I configured NetlifyCMS with Hugo that have markdown files on self-hosted Gitlab. Now I need to make NetlifyCMS preview pane to display data in my Hugo theme style and to render templates and shortcodes correctly.
